Mathlib/Tactic/FunProp.lean

```diff
@@ -127,37 +137,213 @@ displays
--- between different properties
--- between different data
-
-### Debugging
-
+### Type of Theorems
+
+There are four types of theorems that are used a bit differently.
+
+- Lambda Theorems:
+    These theorems are about basic lambda calculus terms like identity, constant, composition, etc.
+    They are used when a bigger expression is decomposed into a sequence of function compositions.
+    They are also used when, for example, we know that a function is continuous in the `x` and `y`
+    variables, but we want continuity only in `x`.
+
+    There are five of them, and they should be formulated in the following way:
+
+    - Identity Theorem
+    ```lean
+    @[fun_prop]
+    theorem continuous_id : Continuous (fun (x : X) => x) := ..
+    ```
+
+    - Constant Theorem
+    ```lean
+    @[fun_prop]
+    theorem continuous_const (y : Y) : Continuous (fun (x : X) => y) := ..
+    ```
+
+    - Composition Theorem
+    ```lean
+    @[fun_prop]
+    theorem continuous_comp (f : Y → Z) (g : X → Y) (hf : Continuous f) (hg : Continuous g) :
+      Continuous (fun (x : X) => f (g x)) := ..
+    ```
+
+    - Apply Theorem
+      It can be either non-dependent version
+    ```lean
+    @[fun_prop]
+    theorem continuous_apply (a : α) : Continuous (fun f : (α → X) => f a) := ..
+    ```
+      or dependent version
+    ```lean
+    @[fun_prop]
+    theorem continuous_apply (a : α) : Continuous (fun f : ((a' : α) → E a') => f a) := ..
+    ```
+
+    - Pi Theorem
+    ```lean
+    @[fun_prop]
+    theorem continuous_pi (f : X → α → Y) (hf : ∀ a, Continuous (f x a)) :
+       Continuous (fun x a => f x a) := ..
+    ```
+
+    Not all of these theorems have to be provided, but at least the identity and composition
+    theorems should be.
+
+    You should also add theorems for `Prod.mk`, `Prod.fst`, and `Prod.snd`. Technically speaking,
+    they fall under the *function theorems* category, but without them, `fun_prop` can't function
+    properly. We are mentioning them as they are used together with *lambda theorems* to break
+    complicated expressions into a composition of simpler functions.
+    ```lean
+    @[fun_prop]
+    theorem continuous_fst (f : X → Y × Z) (hf : Continuous f) :
+        Continuous (fun x => (f x).fst) := ...
+    @[fun_prop]
+    theorem continuous_snd (f : X → Y × Z) (hf : Continuous f) :
+        Continuous (fun x => (f x).snd) := ...
+    @[fun_prop]
+    theorem continuous_prod_mk (f : X → Y) (g : X → Z) (hf : Continuous f) (hg : Continuous g) :
+        Continuous (fun x => (f x, g x)) := ...
+    ```
+
+- Function Theorems:
+    When `fun_prop` breaks complicated expression apart using *lambda theorems* it then uses
+    *function theorems* to prove that each piece is, for example, continuous.
+
+    The function theorem for `Neg.neg` and `Continuous` can be stated as:
+    ```lean
+    @[fun_prop]
+    theorem continuous_neg : Continuous (fun x => Neg.neg x) := ...
```

Comment:
I'm doing it because for example for `a + b` the head of this expression is `HAdd.hAdd`  which is obvious when you write it as `HAdd.hAdd a b` and you don't even know what is the definition of "head of an expression".

Maybe should I switch to some functions that don't have this issue? Maybe `sin` and what would be a good two argument function?

Reply:
No, that's fine, it makes sense, but perhaps a short parenthetical comment to the effect of "(the *head* of an expression is the term which appears first in the tree; for binary operations, like `a + b`, after stripping away the syntactic sugar / notation this becomes `HAdd.hAdd a b`; here we prefer the latter to make the head expression more obvious)"